File: backend/internal/service/spell_service.go
package service

import "rpg-manager/internal/domain"

type spellRepo interface {
	GetAll(edition string) ([]domain.Spell, error)
	GetByCharacter(characterID uint) ([]domain.Spell, error)
	Add(characterID, spellID uint) error
	Remove(characterID, spellID uint) error
}

type SpellService struct{ repo spellRepo }

func NewSpellService(repo spellRepo) *SpellService { return &SpellService{repo: repo} }

func (s *SpellService) GetAll(edition string) ([]domain.Spell, error) {
	return s.repo.GetAll(edition)
}
func (s *SpellService) GetByCharacter(id uint) ([]domain.Spell, error) {
	return s.repo.GetByCharacter(id)
}
func (s *SpellService) Add(charID, spellID uint) error {
	return s.repo.Add(charID, spellID)
}
func (s *SpellService) Remove(charID, spellID uint) error {
	return s.repo.Remove(charID, spellID)
}

File: frontend/src/services/spellService.ts
import api from './api'
import type { Spell } from '../types'

export const spellService = {
  getAll: async (edition?: string): Promise<Spell[]> => {
    const { data } = await api.get('/spells', {
      params: edition ? { edition } : {},
    })
    return data
  },

  getByCharacter: async (characterId: number): Promise<Spell[]> => {
    const { data } = await api.get(`/characters/${characterId}/spells`)
    return data
  },

  add: async (characterId: number, spellId: number): Promise<void> => {
    await api.post(`/characters/${characterId}/spells/${spellId}`)
  },

  remove: async (characterId: number, spellId: number): Promise<void> => {
    await api.delete(`/characters/${characterId}/spells/${spellId}`)
  },
}
